### PR TITLE
Build an API endpoint to retrive all saved reservations

### PR DESCRIPTION
**Pull request by mohamed20163858**
in this pull request I did the following:- 
- I modified the constrain of maximum user length to contain more characters for lengthy names to be 50 chars instead of 16 
- I created an API endpoint to retrieve all already created reservations 
- I created an API endpoint to create new reservations 
- I add 2 new methods called `check_available_dates` and `overlap` to prevent creating faulty overlapping reservations with the ability to provide the reason for prevention with a JSON message format and I tested it with all edge cases and it work perfectly

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -1,7 +1,7 @@
 source 'https://rubygems.org'
 git_source(:github) { |repo| "https://github.com/#{repo}.git" }
 
-ruby '3.2.2'
+ruby '3.1.3'
```
Comment (jmonto55) [inline Gemfile:4]: - [ ] Great work implementing the features for reservations. 💯 However, please kindly ensure to update the Ruby version back to `3.2.2`. 🤓

File: config/database.yml [hunk 2]
```diff
@@ -20,8 +20,8 @@ default: &default
   # For details on connection pooling, see Rails configuration guide
   # https://guides.rubyonrails.org/configuring.html#database-pooling
   pool: <%= ENV.fetch("RAILS_MAX_THREADS") { 5 } %>
-  # username: postgres
-  # password: 2352122Aa.
+  username: postgres
+  password: momo123456
```
Comment (jmonto55) [inline config/database.yml:24]: - [ ] Good work implementing the features. ✔️ However, please kindly ensure to comment your custom username and password. 🤓

**Review by jmonto55 — CHANGES_REQUESTED** [review]
# STATUS: REQUIRED CHANGES 🟡

Hi @mohamed20163858  👋🏽, 

Good job so far! 
There are some issues that you still need to work on to go to the next project but you are almost there!



>  



## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: Gemfile (hunk 1 above)
Comment (mohamed20163858) [reply Gemfile]: thanks alot already did that !

File: config/database.yml (hunk 2 above)
Comment (mohamed20163858) [reply config/database.yml]: thanks a lot already did that

**Review by jmonto55 — COMMENTED** [review]
# STATUS: APPROVED 🟢

Hi @mohamed20163858, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

AgreeSealGIF



## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**